training/src/main/scala/com/airbnb/aerosolve/training/MlpModelTrainer.scala

```diff
@@ -0,0 +1,467 @@
+    val std = trainerOptions.weightInitStd.toFloat
+    val stats = TrainingUtils.getFeatureStatistics(trainerOptions.minCount, pointwise)
+
+    // set up input layer weights
+    var count : Int = 0
+    for (kv <- stats) {
+      val (family, feature) = kv._1
+      if (family != trainerOptions.rankKey) {
+        if (!inputLayerWeights.containsKey(family)) {
+          inputLayerWeights.put(family, new java.util.HashMap[java.lang.String, FloatVector]())
+        }
+        val familyMap = inputLayerWeights.get(family)
+        if (!familyMap.containsKey(feature)) {
+          count = count + 1
+          familyMap.put(feature, FloatVector.getGaussianVector(layerNodeNumber.get(0), std))
+        }
+      }
+    }
+    // set up hidden layer weights
+    for (i <- 0 to numHiddenLayers - 1) {
+      val arr = new java.util.ArrayList[FloatVector]()
+      for (j <- 0 to layerNodeNumber.get(i) - 1) {
+        val fv = FloatVector.getGaussianVector(layerNodeNumber.get(i + 1), std)
+        arr.add(fv)
+      }
+      hiddenLayerWeights.put(i, arr)
+    }
+    // note: bias at each node initialized to zero in this trainer
+    log.info(s"Total number of features is $count")
+    model
+  }
+
+  private def setupUpdateContainer(model: MlpModel) : scala.collection.mutable.HashMap[(String, String), FloatVector] = {
+    val container = scala.collection.mutable.HashMap[(String, String), FloatVector]()
+    // set up input layer weights gradient
+    val inputLayerWeights = model.getInputLayerWeights
+    val n0 = model.getLayerNodeNumber.get(0)
+    for (family <- inputLayerWeights) {
+      for (feature <- family._2) {
+        val key = (family._1, feature._1)
+        container.put(key, new FloatVector(n0))
+      }
+    }
+
+    // set up hidden layer weights gradient
+    val numHiddenLayers = model.getNumHiddenLayers
+    for (i <- 0 until numHiddenLayers) {
+      val thisLayerNodeNum = model.getLayerNodeNumber.get(i)
+      val nextLayerNodeNum = model.getLayerNodeNumber.get(i + 1)
+      for (j <- 0 until thisLayerNodeNum) {
+        val key = (LAYER_PREFIX + i.toString, NODE_PREFIX + j.toString)
+        container.put(key, new FloatVector(nextLayerNodeNum))
+      }
+    }
+
+    // set up bias gradient
+    for (i <- 0 to numHiddenLayers) {
+      // all bias in the same layer are put to the same FloatVector
+      val key = (LAYER_PREFIX + i.toString, BIAS_PREFIX)
+      container.put(key, new FloatVector(model.getLayerNodeNumber.get(i)))
+    }
+
+    container
+  }
+
+  private def computeUpdates(prevUpdate: FloatVector,
+                             momentum: Float,
+                             learningRate: Float,
+                             gradient: FloatVector): FloatVector = {
+    // based on hinton's dropout paper: http://arxiv.org/pdf/1207.0580.pdf
+    val update: FloatVector = new FloatVector(prevUpdate.length)
+    update.multiplyAdd(momentum, prevUpdate)
+    update.multiplyAdd(-(1.0f - momentum) * learningRate, gradient)
```

Comment:
reference to this formula?  I saw one version is mu * v - learning_rate * dx which may transfer to update.multiplyAdd(momentum, prevUpdate)
update.multiplyAdd(-(1.0f ) * learningRate, gradient)